modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java

```diff
@@ -120,105 +142,237 @@
+                            .whenComplete((v, ex) -> {
+                                if (ex != null)
+                                    LOG.error("Exception when closing cursors for " + member.id(), ex);
+                            });
                     }
                 });
             }
-        }
-        else
-            this.metaStorageSvcFut = new CompletableFuture<>();
+        });
+    }
 
-        // TODO: IGNITE-14088: Uncomment and use real serializer factory
-//        Arrays.stream(MetaStorageMessageTypes.values()).forEach(
-//            msgTypeInstance -> net.registerMessageMapper(
-//                msgTypeInstance.msgType(),
-//                new DefaultMessageMapperProvider()
-//            )
-//        );
+    /** {@inheritDoc} */
+    @Override public void beforeNodeStop() {
+        LOG.info("Stopping the MetaStorage manager");
+
+        // cancelling the metaStorageNodeNamesFut will also stop the polling process in metaStorageSvcFut
+        metaStorageNodeNamesFut.cancel(false);
 
-        // TODO: IGNITE-14414 Cluster initialization flow. Here we should complete metaStorageServiceFuture.
-//        clusterNetSvc.messagingService().addMessageHandler((message, senderAddr, correlationId) -> {});
+        try {
+            metaStorageSvcFut.get(10, TimeUnit.SECONDS);
+        }
+        catch (ExecutionException e) {
+            // execution exceptions are not important and are expected to be thrown, e.g. when the component stops
+            // abruptly
+            LOG.debug("Exception during MetaStorageManager stop", e.getCause());
+        }
+        catch (InterruptedException | TimeoutException e) {
+            throw new IgniteInternalException("Exception during MetaStorageManager stop", e);
+        }
     }
 
     /** {@inheritDoc} */
     @Override public void stop() {
-        // TODO: IGNITE-15161 Implement component's stop.
+        // no-op
+    }
+
+    /**
+     * Temporary method for emulating the "init" command.
+     *
+     * @param metastorageNodeNames Names of the nodes hosting the Meta Storage.
+     */
+    // TODO: remove this method: https://issues.apache.org/jira/browse/IGNITE-14414
+    public void init(List<String> metastorageNodeNames) {
+        assert metaStorageSvcFut != null : "MetaStorageManager has not been started";
+
+        metaStorageNodeNamesFut.complete(metastorageNodeNames);
+    }
+
+    /**
+     * Tries to repeatedly retrieve information about the nodes hosting the Meta Storage from a random cluster member.
+     */
+    private CompletableFuture<Collection<String>> joinCluster() {
+        if (metaStorageNodeNamesFut.isDone())
+            return metaStorageNodeNamesFut;
+
+        ClusterNode randomMember = randomMember();
+
+        // if a random member is null, then this is the first node in the cluster and there's nothing to be done
+        // until the "init" command arrives
+        if (randomMember == null)
+            return metaStorageNodeNamesFut;
```

Comment:
Are you sure that this is correct? Can two nodes see this state before connecting to each other or the cluster in general?

Reply:
Good question. I assumed that after the Network Component start, we already have the topology present (i.e. we've joined the cluster). I may be wrong in which case this can be a problem. Need to check...